src/tls_openssl.c

```diff
@@ -245,20 +248,43 @@ const char *cert_errors[] = {
 };
 #undef TLS_ERROR_FIELD
 
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+static OSSL_PROVIDER *providers[2];
+#endif
+
+static void check_providers(xmpp_ctx_t *ctx)
+{
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    if ((providers[0] == NULL) || (providers[1] == NULL)) {
+        strophe_error(ctx, "tls", "could not load all providers");
+    }
+#else
+    (void)ctx;
+#endif
+}
+
 void tls_initialize(void)
 {
 #if OPENSSL_VERSION_NUMBER < 0x10100000L
     SSL_library_init();
     SSL_load_error_strings();
 #else
     OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS, NULL);
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    providers[0] = OSSL_PROVIDER_load(NULL, "legacy");
+    providers[1] = OSSL_PROVIDER_load(NULL, "default");
+#endif
 #endif
     /* init xmppAddr OID */
     _tls_xaddr_nid();
 }
 
 void tls_shutdown(void)
 {
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    OSSL_PROVIDER_unload(providers[1]);
```

Comment:
Does this function accept NULL as an argument? Man page doesn't describe this. Because if "load" operation fails we will call "unload" for NULL unconditionally.